### PR TITLE
Fix #1021: Disable merge identical blocks in the region simplification of canonicalizer pass

### LINKED ISSUE #1021: tco loc("tco_crash.mlir":1:1): error: must be a constant
`tco` is crashed by FIR generated by `bbc` from the following code:

```fortran
      MODULE DMUMPS_SOL_LR
      IMPLICIT NONE

      TYPE BLR_STRUC_T
         INTEGER, DIMENSION(:), POINTER  :: PANELS_L 
         INTEGER, DIMENSION(:), POINTER  :: PANELS_U
         INTEGER, DIMENSION(:), POINTER :: BEGS_BLR_STATIC
      END TYPE BLR_STRUC_T

      TYPE(BLR_STRUC_T), POINTER, DIMENSION(:), SAVE :: BLR_ARRAY

      CONTAINS

      SUBROUTINE DMUMPS_SOL_FWD_LR_SU( IWHDLR, MTYPE )

        INTEGER, INTENT(IN) :: IWHDLR, MTYPE
        INTEGER :: NPARTSASS, NB_BLR

        IF (MTYPE.EQ.1) THEN
          IF ( associated( BLR_ARRAY(IWHDLR)%PANELS_L ) ) THEN
            NPARTSASS = size( BLR_ARRAY(IWHDLR)%PANELS_L )
            NB_BLR = size( BLR_ARRAY(IWHDLR)%BEGS_BLR_STATIC ) - 1
          ENDIF
        ELSE
          IF ( associated( BLR_ARRAY(IWHDLR)%PANELS_U ) ) THEN
            NPARTSASS = size( BLR_ARRAY(IWHDLR)%PANELS_U )
            NB_BLR = size( BLR_ARRAY(IWHDLR)%BEGS_BLR_STATIC ) - 1
          ENDIF
        ENDIF

      END SUBROUTINE DMUMPS_SOL_FWD_LR_SU 

      END MODULE DMUMPS_SOL_LR
```

Here's the stack dump

```
tco ./tco_crash.mlir 
loc("./tco_crash.mlir":1:1): error: must be a constant
LLVM ERROR: aborting
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.  Program arguments: /home/ryuta/packages/llvm/flang/build/bin/tco ./tco_crash.mlir
 #0 0x00007f9bc6de53c2 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/ryuta/packages/llvm/flang/f18-llvm-project/llvm/lib/Support/Unix/Signals.inc:565:11
 #1 0x00007f9bc6de556b PrintStackTraceSignalHandler(void*) /home/ryuta/packages/llvm/flang/f18-llvm-project/llvm/lib/Support/Unix/Signals.inc:632:1
 #2 0x00007f9bc6de3b97 llvm::sys::RunSignalHandlers() /home/ryuta/packages/llvm/flang/f18-llvm-project/llvm/lib/Support/Signals.cpp:76:5
 #3 0x00007f9bc6de63a3 SignalHandler(int) /home/ryuta/packages/llvm/flang/f18-llvm-project/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007f9bccd00630 __restore_rt sigaction.c:0:0
 #5 0x00007f9bc6276387 raise (/lib64/libc.so.6+0x36387)
 #6 0x00007f9bc6277a78 abort (/lib64/libc.so.6+0x37a78)
 #7 0x00007f9bc6cb0f71 (/home/ryuta/packages/llvm/flang/build/lib/libLLVMSupport.so.13git+0x22ff71)
 #8 0x00007f9bc6cb0e22 (/home/ryuta/packages/llvm/flang/build/lib/libLLVMSupport.so.13git+0x22fe22)
 #9 0x00007f9bccc2917c (/home/ryuta/packages/llvm/flang/build/lib/libFIRCodeGen.so.13git+0x16117c)
#10 0x00007f9bccc4ce62 (anonymous namespace)::CoordinateOpConversion::getIntValue(mlir::Value) const /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/lib/Optimizer/CodeGen/CodeGen.cpp:2273:5
#11 0x00007f9bccc4c1bf (anonymous namespace)::CoordinateOpConversion::getFieldNumber(fir::RecordType, mlir::Value) const /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/lib/Optimizer/CodeGen/CodeGen.cpp:2185:5
#12 0x00007f9bccc4c6bb (anonymous namespace)::CoordinateOpConversion::validCoordinate(mlir::Type, llvm::ArrayRef<mlir::Value>) const /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/lib/Optimizer/CodeGen/CodeGen.cpp:2234:22
#13 0x00007f9bccc4a2e8 (anonymous namespace)::CoordinateOpConversion::doRewrite(fir::CoordinateOp, mlir::Type, llvm::ArrayRef<mlir::Value>, mlir::ConversionPatternRewriter&) const /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/lib/Optimizer/CodeGen/CodeGen.cpp:2096:9
#14 0x00007f9bccc497eb (anonymous namespace)::FIROpAndTypeConversion<fir::CoordinateOp>::matchAndRewrite(fir::CoordinateOp, llvm::ArrayRef<mlir::Value>, mlir::ConversionPatternRewriter&) const /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/lib/Optimizer/CodeGen/CodeGen.cpp:278:12
#15 0x00007f9bccc49717 mlir::detail::OpOrInterfaceConversionPatternBase<fir::CoordinateOp>::matchAndRewrite(mlir::Operation*, llvm::ArrayRef<mlir::Value>, mlir::ConversionPatternRewriter&) const /home/ryuta/packages/llvm/flang/f18-llvm-project/llvm/../mlir/include/mlir/Transforms/DialectConversion.h:386:12
#16 0x00007f9bc8c5aee4 mlir::ConversionPattern::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:1525:10
#17 0x00007f9bc8aaea9d mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Rewrite/PatternApplicator.cpp:201:25
#18 0x00007f9bc8c7f3ff (anonymous namespace)::OperationLegalizer::legalizeWithPattern(mlir::Operation*, mlir::ConversionPatternRewriter&) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:1801:21
#19 0x00007f9bc8c7ea70 (anonymous namespace)::OperationLegalizer::legalize(mlir::Operation*, mlir::ConversionPatternRewriter&) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:1718:17
#20 0x00007f9bc8c7dedc (anonymous namespace)::OperationConverter::convert(mlir::ConversionPatternRewriter&, mlir::Operation*) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:2213:26
#21 0x00007f9bc8c5e697 (anonymous namespace)::OperationConverter::convertOperations(llvm::ArrayRef<mlir::Operation*>) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:2258:16
#22 0x00007f9bc8c5e8fb mlir::applyFullConversion(llvm::ArrayRef<mlir::Operation*>, mlir::ConversionTarget&, mlir::FrozenRewritePatternSet const&) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:2819:22
#23 0x00007f9bc8c5e95a mlir::applyFullConversion(mlir::Operation*, mlir::ConversionTarget&, mlir::FrozenRewritePatternSet const&) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Transforms/Utils/DialectConversion.cpp:2824:10
#24 0x00007f9bccbea096 (anonymous namespace)::FIRToLLVMLowering::runOnOperation() /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/lib/Optimizer/CodeGen/CodeGen.cpp:3112:22
#25 0x00007f9bc87a177e mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Pass/Pass.cpp:385:21
#26 0x00007f9bc87a1e82 mlir::detail::OpToOpPassAdaptor::runPipeline(llvm::iterator_range<llvm::pointee_iterator<std::__1::unique_ptr<mlir::Pass, std::__1::default_delete<mlir::Pass> >*, mlir::Pass> >, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Pass/Pass.cpp:426:16
#27 0x00007f9bc87a3ecb mlir::PassManager::runPasses(mlir::Operation*, mlir::AnalysisManager) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Pass/Pass.cpp:692:10
#28 0x00007f9bc87a3b8c mlir::PassManager::run(mlir::Operation*) /home/ryuta/packages/llvm/flang/f18-llvm-project/mlir/lib/Pass/Pass.cpp:672:60
#29 0x000000000024d954 compileFIR(mlir::PassPipelineCLParser const&) /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/tools/tco/tco.cpp:140:26
#30 0x000000000024d1c6 main /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/tools/tco/tco.cpp:165:23
#31 0x00007f9bc6262555 __libc_start_main (/lib64/libc.so.6+0x22555)
#32 0x000000000024c399 _start (/home/ryuta/packages/llvm/flang/build/bin/tco+0x24c399)
Aborted
```

```
head -10 ./tco_crash.mlir 
func @_QMdmumps_sol_lrPdmumps_sol_fwd_lr_su(%arg0: !fir.ref<i32>, %arg1: !fir.ref<i32>) {
  %c0_i64 = constant 0 : i64
  %c0 = constant 0 : index
  %c1_i32 = constant 1 : i32
  %0 = fir.address_of(@_QMdmumps_sol_lrEblr_array) : !fir.ref<!fir.box<!fir.ptr<!fir.array<?x!fir.type<_QMdmumps_sol_lrTblr_struc_t{panels_l:!fir.box<!fir.ptr<!fir.array<?xi32>>>,panels_u:!fir.box<!fir.ptr<!fir.array<?xi32>>>,begs_blr_static:!fir.box<!fir.ptr<!fir.array<?xi32>>>}>>>>>
  %1 = fir.alloca i32 {bindc_name = "nb_blr", uniq_name = "_QMdmumps_sol_lrFdmumps_sol_fwd_lr_suEnb_blr"}
  %2 = fir.alloca i32 {bindc_name = "npartsass", uniq_name = "_QMdmumps_sol_lrFdmumps_sol_fwd_lr_suEnpartsass"}
  %3 = fir.load %arg1 : !fir.ref<i32>
  %4 = cmpi eq, %3, %c1_i32 : i32
  cond_br %4, ^bb1, ^bb3
  ...
```

### PR DESCRIPTION
**Pull request by clementval**
This is a quick fix for issue #1021. As mentioned in the issue, the problem comes from a merge of the two identical blocks that introduces a block argument for the `fir.field_index`. After that the `fir.field_index` is not directly visible in the translation of `fir.coordinate_of` and `tco` fails.
A longer term fix would be to update the algorithm in `mergeIdenticalBlocks` to take into account operation that cannot be moved to block argument. This would probably require `fir.field_index` to adopt a specific trait like `ConstantLike` or smth similar in order to be detected by the algorithm.

The two first commits are cherry-pick from upstream and they add the initial configuration for the canonicalizer pass. 

fe0f6fb and 0fd45ae will be post on phabricator as they can be upstreamed directly. 
- https://reviews.llvm.org/D109579
- https://reviews.llvm.org/D109580

The two following tests have been updated since the merge is not done anymore. 
- `flang/test/Lower/format-statement.f90`
- `flang/test/Fir/loop01.fir`

### CODE REVIEW

**Review by jeanPerier — APPROVED** [review]
LGTM

File: flang/test/Lower/identical-block-merge-disable.f90 [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  ENDIF
+
+END SUBROUTINE DMUMPS_SOL_FWD_LR_SU 
+
+END MODULE DMUMPS_SOL_LR
+
+! CHECK: func @_QMdmumps_sol_lrPdmumps_sol_fwd_lr_su
+! CHECK: ^bb[[BLK1:[1-9]]]
+! CHECK: cond_br %{{.*}}, ^bb[[BLKTRUE:[1-9]]], ^bb[[BLKFALSE:[1-9]]]
+! CHECK: ^bb[[BLKTRUE]]: // pred: ^bb[[BLK1]]
```
Comment (schweitzpgi) [inline flang/test/Lower/identical-block-merge-disable.f90:42]: I think we want to error on the side of very restrictive tests before upstreaming (so review suggestions that break the compiler can be quickly caught). Consider just running the bbc output through mlir/utils/generate-test-checks.py and slurping in that as the test.

**Review by schweitzpgi — APPROVED** [review]

File: flang/test/Lower/identical-block-merge-disable.f90 (hunk 1 above)
Comment (clementval) [reply flang/test/Lower/identical-block-merge-disable.f90]: Updated the test with the auto-generated checks.

**Comment by sscalpone** [comment]
Is this issue unique to the fir dialect?

**Comment by clementval** [comment]
> Is this issue unique to the fir dialect?

So far we are the only dialect to have raised this issue but I think this could happen to other dialect as well.

**Comment by clementval** [comment]
Waiting on more feedback from https://reviews.llvm.org/D109579 to move forward.

**Comment by clementval** [comment]
This solution is not acceptable on the MLIR side.

**Pull request closed without merging**